src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,507 @@
+  }
+
+  /// Return the number of resources in this set.
+  size_t Size() const {
+    size_t size = custom_resources_.size();
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      if (predefined_resources_[i] != 0) {
+        size++;
+      }
+    }
+    return size;
+  }
+
+  /// Return true if this set is empty.
+  bool IsEmpty() const { return Size() == 0; }
+
+  /// Return a set that contains all resource ids in this set.
+  absl::flat_hash_set<ResourceID> ResourceIds() const {
+    absl::flat_hash_set<ResourceID> res;
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] != 0) {
+        res.insert(ResourceID(i));
+      }
+    }
+    for (auto &entry : custom_resources_) {
+      res.insert(ResourceID(entry.first));
+    }
+    return res;
+  }
+
+  /// Return a map from the resource ids to the values.
+  absl::flat_hash_map<ResourceID, FixedPoint> ToMap() const {
+    absl::flat_hash_map<ResourceID, FixedPoint> res;
+    for (auto resource_id : ResourceIds()) {
+      res.emplace(resource_id, Get(resource_id));
+    }
+    return res;
+  }
+
+  ResourceRequest &operator=(const ResourceRequest &other) {
+    this->predefined_resources_ = other.predefined_resources_;
+    this->custom_resources_ = other.custom_resources_;
+    this->requires_object_store_memory_ = other.requires_object_store_memory_;
+    return *this;
+  }
+
+  ResourceRequest operator+(const ResourceRequest &other) {
+    ResourceRequest res = *this;
+    res += other;
+    return res;
+  }
+
+  ResourceRequest operator-(const ResourceRequest &other) {
+    ResourceRequest res = *this;
+    res -= other;
+    return res;
+  }
+
+  ResourceRequest &operator+=(const ResourceRequest &other) {
+    auto resource_ids = ResourceIds();
+    for (auto resource_id : resource_ids) {
+      Set(resource_id, Get(resource_id) + other.GetOrZero(resource_id));
+    }
+    for (auto &resource_id : other.ResourceIds()) {
+      if (resource_ids.find(resource_id) == resource_ids.end()) {
+        Set(resource_id, other.Get(resource_id));
+      }
+    }
+    return *this;
+  }
+
+  ResourceRequest &operator-=(const ResourceRequest &other) {
+    auto resource_ids = ResourceIds();
+    for (auto resource_id : resource_ids) {
+      Set(resource_id, Get(resource_id) - other.GetOrZero(resource_id));
+    }
+    for (auto &resource_id : other.ResourceIds()) {
+      if (resource_ids.find(resource_id) == resource_ids.end()) {
+        Set(resource_id, -other.Get(resource_id));
+      }
+    }
+    return *this;
+  }
+
+  bool operator==(const ResourceRequest &other) const {
+    return FixedPointVectorEqual(predefined_resources_, other.predefined_resources_) &&
+           this->custom_resources_ == other.custom_resources_;
+  }
+
+  bool operator!=(const ResourceRequest &other) const { return !(*this == other); }
+
+  /// Check whether this set is a subset of another one.
+  bool operator<=(const ResourceRequest &other) const {
```

Comment:
Should we just use `IsSubset` instead of `<=`?

Reply:
I thought of "IsSubset". But changed to "<=" as "a <= b" looks clearer to me. No super strong preference thought.